execution/src/sce_ledger.rs

```diff
@@ -0,0 +1,332 @@
+use std::sync::{Arc, Mutex};
+
+use crate::ExecutionError;
+use crypto::hash::Hash;
+use models::ModelsError;
+use models::{address::AddressHashMap, hhasher::HHashMap, Address, Amount, AMOUNT_ZERO};
+use wasmer::Module;
+
+/// an entry in the SCE ledger
+#[derive(Debug, Clone, Default)]
+pub struct SCELedgerEntry {
+    pub balance: Amount,
+    pub opt_module: Option<Module>,
+    pub data: HHashMap<Hash, Vec<u8>>,
+}
+
+impl SCELedgerEntry {
+    /// applies an entry update to self
+    pub fn apply_entry_update(&mut self, update: &SCELedgerEntryUpdate) {
+        // balance
+        if let Some(new_balance) = update.update_balance {
+            self.balance = new_balance;
+        }
+
+        // module
+        if let Some(opt_module) = &update.update_opt_module {
+            self.opt_module = opt_module.clone();
+        }
+
+        // data
+        for (data_key, data_update) in update.update_data.iter() {
+            match data_update {
+                Some(new_data) => {
+                    self.data.insert(*data_key, new_data.clone());
+                }
+                None => {
+                    self.data.remove(data_key);
+                }
+            }
+        }
+    }
+}
+
+// optional updates to be applied to a ledger entry
+#[derive(Debug, Clone)]
+pub struct SCELedgerEntryUpdate {
+    pub update_balance: Option<Amount>,
+    pub update_opt_module: Option<Option<Module>>,
```

Comment:
Every field in SCELedgerEntryUpdate is an option: each of the the values is either updated (Some) or not updated (None). 

But update_opt_module is already an Option because an address may or may not have a Module. Since removing or adding a module from an address is a possible SCE ledger entry update, we must have an Option<Option<Module>>

Reply:
seems like a weird design to me

Reply:
I don't find it particularly shocking because it is consistent with all other fields that you thought were OK, it just happens that the type of this field in particular is itself already an Option. But feel free to propose an alternative if it's consistent with other fields.